DDF-2815: Add UTM coordinate support to CSW endpoint

troymohl opened the pull request:
#### What does this PR do?
Adds UTM coordinate support for OGC Filters in the CSW endpoint.

#### Who is reviewing it (please choose AT LEAST two reviewers that need to approve the PR before it can get merged)?
@bdeining 
@rzwiefel 
@ricklarsen 

#### Select at least one member from relevant component team(s) from below (at least one component team member needs to approve the PR).
[OGC](https://github.com/orgs/codice/teams/OGC)

#### Choose 2 committers to review/merge the PR (please choose ONLY two committers from below, delete the rest).
@kcwire
@millerw8

#### How should this be tested? (List steps with links to updated documentation)
Build & Run Unit Tests and Itests

With postman, POST the following to the CSW endpoint and verify no exceptions are returned:
`https://localhost:8993/services/csw`
`<?xml version="1.0" ?>
<GetRecords xmlns="http://www.opengis.net/cat/csw/2.0.2"
        xmlns:ogc="http://www.opengis.net/ogc"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xmlns:gml="http://www.opengis.net/gml"
        service="CSW"
        version="2.0.2"
        maxRecords="4"
        startPosition="1"
        resultType="results"
        outputFormat="application/xml"
        outputSchema="http://www.opengis.net/cat/csw/2.0.2"
        xsi:schemaLocation="http://www.opengis.net/cat/csw/2.0.2 ../../../csw/2.0.2/CSW-discovery.xsd">
    <Query typeNames="Record">
        <ElementSetName>summary</ElementSetName>
        <Constraint version="1.1.0">
            <ogc:Filter>
                <ogc:Intersects>
                    <ogc:PropertyName>ows:BoundingBox</ogc:PropertyName>
                    <gml:Envelope srsName="urn:ogc:def:crs:EPSG:32636">
                        <gml:lowerCorner>171070 1106907</gml:lowerCorner>
                        <gml:upperCorner>225928 1106910</gml:upperCorner>
                    </gml:Envelope>
                </ogc:Intersects>
            </ogc:Filter>
        </Constraint>
    </Query>
</GetRecords>`

#### Any background context you want to provide?
UTM Coordinates are handled with the following coordinate reference systems:
EPSG:326XX -> North UTM coordinates where XX is the UTM zone
EPSG:327XX -> South UTM coordinates where XX is the UTM zone

#### What are the relevant tickets?
https://codice.atlassian.net/browse/DDF-2815

#### Screenshots (if appropriate)
#### Checklist:
- [X] Documentation Updated
- [X] Change Log Updated
- [X] Update / Add Unit Tests
- [X] Update / Add Integration Tests

bdeining commented on catalog/spatial/csw/spatial-csw-endpoint/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/endpoint/mappings/CswRecordMapperFilterVisitor.java:
```diff
@@ -503,5 +578,21 @@ public Object visit(Function function, Object extraData) {
         }
     }
+
+    private static void convertGeometryExpression(Expression expression) {
```
nit : should this be renamed to convertGeometryExpressionToEPSG4326?

troymohl replied:
Updated with the more descriptive name.

kcwire commented:
@troymohl I think this warrants an update to the change log. 

@ricklarsen Can you remind me where we have the guidelines for when and how to update the change log?

kcwire commented on distribution/docs/src/main/resources/_contents/_endpoints/csw-endpoint-contents.adoc:
```diff
@@ -1230,6 +1230,43 @@ It is possible to receive a response to a `GetRecords` query that conforms to th
 
+.`GetRecords` XML Request using UTM Coordinates
+UTM coordinates can be used when making a CSW GetRecords request using an `ogc:Filter`.
+UTM coordinates should use EPSG:326XX where XX is the zone within the northern hemisphere.
```
I think we want to be more explicit here. The `EPSG:326XX` is the srsName.  Do we also require the exact urn that is in the sample below? Just want to remove any ambiguities.

troymohl replied:
The full URN wasn't required, so I updated the doc and test to reflect that.

kcwire commented on libs/geospatial/src/main/java/org/codice/ddf/libs/geo/util/GeospatialUtil.java:
```diff
@@ -13,9 +13,34 @@
+
+import com.vividsolutions.jts.geom.Geometry;
 
 public class GeospatialUtil {
```
Can we get some glass level javadoc?

kcwire commented on libs/geospatial/src/main/java/org/codice/ddf/libs/geo/util/GeospatialUtil.java:
```diff
@@ -171,4 +196,73 @@ public static Double parseDMSLongitudeWithDecimalSeconds(String dmsLon)
+        } catch (FactoryException | TransformException e) {
+            LOGGER.debug("Unable to convert {} into {}", sourceCrsName, EPSG_4326, e);
+        }
+        return null;
```
Should we throw instead of returning null? how would a caller know what went wrong without the exception?

troymohl replied:
Updated these methods to throw an exception rather than just returning null.

kcwire commented on libs/geospatial/src/main/java/org/codice/ddf/libs/geo/util/GeospatialUtil.java:
```diff
@@ -171,4 +196,73 @@ public static Double parseDMSLongitudeWithDecimalSeconds(String dmsLon)
+                LOGGER.debug("Converted CRS {} into {} : {}", sourceCrs, EPSG_4326, geometry);
+            }
+        } catch (FactoryException | TransformException e) {
+            LOGGER.debug("Unable to convert {} into {}", sourceCrs, EPSG_4326, e);
```
It appears we are hiding a failure case to the caller by not throwing this exception up.

troymohl replied:
Updated these methods to throw an exception rather than just returning null.

jlcsmith commented:
@troymohl - it looks like you updated the Wfs source to transform geometries to 4236 in the [AbstractFeatureConverter](https://github.com/codice/ddf/pull/1712/files#diff-5e03baa5c2a95c8f23ea95c99d26ce05R288).  However, you didn't make a similar change on the Csw Source side.  Any reason for this?

troymohl commented:
@jlcsmith I really just extracted the Wfs code into our common GeospatialUtil class and added code for the CSW Endpoint to call to do a similar conversion. I also extracted some statics that would be useful in both methods. I left the existing WFS code (and CSW) intact except for this minor refactor of a method that wasn't necessarily WFS specific.

jlcsmith commented:
@troymohl - thanks for the explanation.  That makes sense.  I think we should make a similar change in the Csw Source but that sounds like a separate ticket

ricklarsen commented:
@kcwire We don't really have formal guidelines for the changelog, but I'll get some up on the documentation style guide wiki page.

troymohl commented:
@kcwire @ricklarsen I have updated the CHANGES.md file with this PR to add a 2.11 section and add my ticket in the `New Features` section

millerw8 commented:
Can we add to this documentation that we have UTM support when using an OGC Filter but the CQL BNF doesn't support it: 
https://github.com/troymohl/ddf/blob/87d7a409173458bac2c61ceddd9cd5a8c6273fbd/distribution/docs/src/main/resources/_contents/_endpoints/csw-endpoint-contents.adoc

rzwiefel commented on catalog/spatial/csw/spatial-csw-endpoint/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/endpoint/mappings/CswRecordMapperFilterVisitor.java:
```diff
@@ -503,5 +579,26 @@ public Object visit(Function function, Object extraData) {
     }
+
+    private static void convertGeometryExpressionToEpsg4326(Expression expression) {
+        if (expression instanceof LiteralExpressionImpl) {
```
can't tell if this is better or worse. less complexity yes, but also i dont like the java not `!`. Anyways just posting it for comparison. No changes necessary
```java
if (!(expression instanceof LiteralExpressionImpl)) {
    return;
}

LiteralExpressionImpl literalExpression = (LiteralExpressionImpl) expression;
Object valueObj = literalExpression.getValue();
if (!(valueObj instanceof Geometry)) {
    return;
} 

Geometry geometry = (Geometry) valueObj;
Object userDataObj = geometry.getUserData();
if (!(userDataObj instanceof CoordinateReferenceSystem)) {
    return;
}

CoordinateReferenceSystem sourceCRS = (CoordinateReferenceSystem) userDataObj;
try {
    Geometry convertedGeometry = GeospatialUtil.transformToEPSG4326LonLatFormat(geometry, sourceCRS);
    literalExpression.setValue(convertedGeometry);
} catch (GeoFormatException e) {
    LOGGER.trace("Unable to convert geometry to EPSG:4326 format", e);
}
    
```

bdeining commented on libs/geospatial/src/main/java/org/codice/ddf/libs/geo/util/GeospatialUtil.java:
```diff
@@ -171,4 +199,75 @@ public static Double parseDMSLongitudeWithDecimalSeconds(String dmsLon)
+    public static Geometry transformToEPSG4326LonLatFormat(Geometry geometry, String sourceCrsName)
+            throws GeoFormatException {
+        try {
+            CoordinateReferenceSystem sourceCrs = CRS.decode(sourceCrsName);
```
sourceCrsName and geometry could be null here and throw an NPE
